Постановка целей

InfernalBeard opened the pull request:
https://faang-school.atlassian.net/browse/BJS2-40568
Комментарии внутри кода

Commit:
feat(controller): create GoalController with createGoal endpoint to handle user goal creation requests

Commit:
feat(service): add GoalService with createGoal method to handle goal creation business logic

Commit:
feat(service): add skill existence validation in GoalService createGoal method

Commit:
refactor(controller): refactor createGoal method in GoalController

Commit:
feat(controller): Implement updateGoal method in GoalController

Commit:
feat(dto): Implement GoalDto class with fields

Commit:
feat(service): Implement updateGoal method in GoalService

Commit:
feat(service): Implement updateGoal method in GoalService and GoalController

Commit:
feat(service): Add deleteGoal method for deleting goals.
Implemented deleteGoal method in GoalService to delete a goal by its ID.
Added validation to check if the goal exists before deletion.

Commit:
feat(controller): Implement deleteGoal method in GoalController.
Added deleteGoal method to handle DELETE requests for deleting a goal.
Returned appropriate response codes for successful and unsuccessful deletion.

Commit:
feat(goalfilterdto): create GoalFilterDto class with fields

Commit:
feat(goaldto): upgrade field status

Commit:
feat(controller)Add GoalController with endpoint for getting goals by user with filters

Commit:
feat(dto): Implement GoalService with findGoalsByUserId method and filtering logic based on GoalFilterDto

Commit:
feat(dto): some question for realisation

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,80 @@
+
+        return createdGoal.map(ResponseEntity::ok)
+                .orElseGet(() -> ResponseEntity.badRequest().build());
+    }
```
Я бы предложил во всех методах контроллера использовать аннотацию `@ResponseStatus` вместо `ResponseEntity`, это позволит избавиться от лишней логики обработки этих ответов в контроллере. Обработку ошибок можно добавить в сервисе, где потом также можно будет создать все необходимые обработки ошибок, которые также смогут отправлять ответы типа `BadRequest`. Например можно рассмотреть аннотацию @RestControllerAdvice.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<Goal> createGoal(@RequestParam Long userId, @RequestBody Goal goal, @RequestParam List<Long> skillIds) {
```
Если смотреть по заданию то на вход должен приходить Goal, но мне кажется, что все таки там должен быть DTO, потому что пользователь не может/не должен отправлять конкретную сущность в контроллер, а может/должен отправлять данные которые необходимые для создания его цели

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,80 @@
+
+    @GetMapping("/{goalId}/subtasks")
+    public List<GoalDto> findSubtasksByGoalId(@PathVariable long goalId, @RequestParam(required = false) String statusFilter) {
+        // Вызываем метод из GoalService, передавая идентификатор цели и опциональный фильтр
```
я предлжил бы убирать комменты перед пушем и `goalId` должен быть `Long`

DaurenBerdibekov commented on src/main/java/school/faang/user_service/dto/goal/GoalDto.java:
```diff
@@ -0,0 +1,20 @@
+    private String title;
+    private GoalStatus status;
+    private List<Long> skillIds;
+}
```
Над полями я бы повесил аннотацию @NotNull, что бы поля были провалидированы.

EkaterinaBadratdinova replied:
1. Мне кажется, когда пользователь создает Цель, то он изначально может назначить (выбрать из списка на фронте) кто будет ментором, т.е. можно дополнить полем private Long mentorId.
2. У каждой цели должен быть дэдлайн (иначе что это за цель))), поэтому я бы у ДТО добавила бы поле private LocalDateTime deadline, но сделала бы его не обязательным.
Если рассуждать еще дальше, то эта ДТОшка у нас является еще и ответом на запрос, т.е. после сохранения в БД нам возвращается сущность со всеми полями и на основе этих данных будет отрисовываться фронт. И фронт при отрисовке не получит (в твоем случае) данные ни о менторе, ни о дэдлайне. Иными словами, у тебя и БД сушность цели будет лежать без назначенного ментора и дэдлайна (навечно), а у сущности то эти поля есть.
3. Я бы убрала статус цели из ДТО. Как и писала, первое - представь как ты на фронте создавал бы цель. Ты мог бы сам ей назначить статус ACTIVE или COMPLETED? По твоей логике Да. Но это делает программа, а пользователь просто нажимает кнопочку - Создать, Завершить, а программа сама назначает эти поля. Следовательно, перенести логику назначения этого поля в сервис.

Рекомендация: если сомневаешься, какие поля создавать у ДТО, то посмотри какие поля есть у сущности и определи, что было бы критически важно для пользователя, для дальнейшего отображения после создания цели и т.д.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+    public GoalService(GoalRepository goalRepository, SkillRepository skillRepository) {
+        this.goalRepository = goalRepository;
+        this.skillRepository = skillRepository;
+    }
```
Вместо явного конструктора я бы предложил использовать `@RequiredArgsConstructor`. Также, над классом я бы повесил `@Slf4j` и логгировал некоторые операции.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        Goal createdGoal = goalRepository.create(goal.getTitle(), goal.getDescription(), goal.getParent().getId());
+        return Optional.of(createdGoal);
```
я предложил бы использовать `Optional.ofNullable()`, так как это поможет избежать выброса исключения, если объект будет null. Но если точно известно что там не будет null, то `Optional.of()` будет достаточно.

DaurenBerdibekov replied:
Вообще я не уверен что метод `createGoal` должен быть `Optional`, потому что он записывает данные в бд. То есть если создание цели происходит успешно, то все проверки и условия прошли успешно, поэтому цель полюбому должна сохраняться в бд и возвращаться вместе с сгенерированным айди.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+        Goal existingGoal = goalRepository.findById(goalId).orElseThrow(() ->
+                new IllegalArgumentException("Goal not found."));
+
+        if (existingGoal.getStatus() == GoalStatus.COMPLETED) {
```
вместо `==` я бы использовал `equals()`. Сравнением знаками равно не совсем подходят для сравнения строк.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        for (Long skillId : goalDto.getSkillIds()) {
+            if (!skillRepository.existsById(skillId)) {
```
В данный момент, количество обращений в бд для проверки одного `id` зависит от количества содержащихся айдишников в `goalDto`. Возможно стоит подумать как этого можно избежать. Мог бы предложить получать все айдишники скилов и записывать их в лист. И уже в цикле проходится по листу. Наверное не лучшее решение, но это позволит снять нагрузку с бд и ускорит программу.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+        existingGoal.setDescription(goalDto.getDescription());
+        existingGoal.setStatus(GoalStatus.valueOf(String.valueOf(goalDto.getStatus())));
+
+        if (GoalStatus.COMPLETED == GoalStatus.valueOf(String.valueOf(goalDto.getStatus()))) {
```
тут тоже поменял бы на `equals()`

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+            for (User user : users) {
+                for (Long skillId : goalDto.getSkillIds()) {
+                    skillRepository.assignSkillToUser(skillId, user.getId());
+                }
```
Здесь тоже я бы предложил подумать как избавиться от вложенного цикла, так как это не совсем оптимально. Но это я просто для информации.

DaurenBerdibekov replied:
Это больше к теме `Time Complexity` если будет интересно

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+
+    public List<GoalDto> findSubtasksByGoalId(long goalId, String statusFilter) {
+        // Получаем поток подзадач по идентификатору родительской цели
+        Stream<Goal> subtasksStream = goalRepository.findByParent(goalId);
```
я предложил бы сразу записывать в лист, так как стрим не хранит данные в себе.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+        // Фильтруем подзадачи по статусу, если фильтр передан
+        if (statusFilter != null && !statusFilter.isEmpty()) {
+            subtasksStream = subtasksStream.filter(goal -> goal.getStatus().toString().equalsIgnoreCase(statusFilter));
+        }
```
я бы предложил весь метод реализовать одним стримом и вернуть лист

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+                goal.getStatus() != null ? goal.getStatus().toString() : null, // Преобразуем статус в строку
+                skillIds // Добавляем список идентификаторов навыков
+        );
+    }
```
Тут я бы попробовал реализовать это через MapStruct

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,148 @@
+                        goal.getSkillIds() //нужно ли делать фильтр по этим полям?
+                ))
+                .collect(Collectors.toList());
+    }
```
этот метод я бы тоже предложил сделать одним стримом

Commit:
feat: update GoalDto with fields for goal details and related attributes

Commit:
feat: update GoalFilterDto for filtering goals based on criteria

Commit:
feat: Add GoalMapper for mapping between Goal and GoalDto

Commit:
feat: Add GoalFilter interface for flexible goal filtering

Commit:
feat: Implement GoalDescriptionFilter for description-based goal filtering

Commit:
feat: Implement GoalStatusFilter for description-based goal filtering

Commit:
feat: add methods GoalStatusFilter for description-based goal filtering

Commit:
feat: Add GoalTitleFilter to filter goals based on title

Commit:
feat: Add GoalValidator with goal validation methods for user goals and goal status checks

Commit:
feat: Add custom DataValidationException for data validation errors

Commit:
feat: Add transactional method to fetch user by ID with exception handling for non-existing user

Commit:
feat: upgrade UserService

Commit:
refactor: Update GoalService

Commit:
refactor: Update oalController with CRUD and filtering endpoints for goals management

Commit:
refactor: Update GoalService

Commit:
feat: Add SkillValidator to validate all skill IDs exist in the database

Commit:
refactor: Add UserService class for Lombok's Slf4j and Spring's Service annotations to validate all skill IDs exist in the database

Commit:
feat(service): Add SkillValidator dependency to GoalService for skill validation

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+
+    @PostMapping("/users/{userId}")
+    @ResponseStatus(HttpStatus.CREATED)
+    public GoalDto createGoal(@PathVariable("userId") @Positive long userId,
```
В контроллере входные параметры объявляются как `long`, а в сервисе `Long`. Желательно использовать чтото одно. Но в данном рекомендуется использовать `Long`, так как это объектный тип и он будет передавать `null`, вместо `0` как в `long`. Потому что айди цели будет сгенерированно только в бд, а до этого айди будет неизвестным, и поэтому было бы лучше сразу обозначить, что айди является `null`, а не дефолтное значение `0` как было бы в `long`

InfernalBeard replied:
Согласен! Чтобы код был более согласованным и гибким, можно использовать Long и в контроллере, и в сервисе

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+                                              @RequestParam(value = "filterDescription", required = false) String filterDescription,
+                                              @RequestParam(value = "filterStatus", required = false) GoalStatus filterStatus,
+                                              @RequestParam(value = "filterSkills", required = false) List<Long> filterSkills) {
+        GoalFilterDto filterDto = new GoalFilterDto(filterTitle, filterDescription, filterStatus, filterSkills);
```
тут я бы предложил сразу добавлять дто в метод, вместо создания отдельного объекта

InfernalBeard replied:
Так же изменил и у второго метода.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,65 @@
+@RestController
+@RequestMapping("/v1/goals")
+@AllArgsConstructor
+public class GoalController {
```
я бы еще добавил в контроллер @Slf4j и добалял логгинг в методы

InfernalBeard replied:
И опять согласен! Сделано!

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+    private final UserService userService;
+    private final SkillValidator skillValidator;
+
+    @Transactional
```
как выяснилось из последнего митапа, аннотацию @Transactional желательно вешать над методами, выполнение которых является критичным. То есть нам прям важно чтобы данные записались либо все откатилось назад. Поэтому нужно смотреть что мы хотим.

zaqqaz226 replied:
Тоже не вижу тут смысла использовать транзакцию.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/validator/goal/GoalValidator.java:
```diff
@@ -0,0 +1,70 @@
+        if (goalId == null) {
+            throw new ValidationException("Goal id can't be null");
+        }
+        if (goalId < 0) {
```
не уверен что такая проверка нужна, так как `id` для каждого Goal присвается базой данных и этот `id` уже оп умолчанию не сможет быть менньше `1`. А до присваения `id` базой данных значение `id` будет `null`, так как мы используем `Long`. Даже если такая проверка была бы нужно, я бы поставил `goalId <= 0` так как айди не может быть 0.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/validator/goal/GoalValidator.java:
```diff
@@ -0,0 +1,70 @@
+@RequiredArgsConstructor
+public class GoalValidator {
+
+    private static final int MAX_USER_GOALS_LIMIT = 3;
```
предложил бы вынести значение в конфиг файл

EkaterinaBadratdinova replied:
Рекомендовала бы вынести переменную в конфиг и внедрять в класс значение из конфига, используя аннотацию Value

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Transactional
+    public List<GoalDto> findSubtasksByGoalId(long goalId, GoalFilterDto filterDto) {
+        Stream<Goal> goals = goalRepository.findByParent(goalId);
```
я бы предложил сохранять реультат сразу в лист и поменять входной параметр в методе `filterGoals` на `List<Goal>`

Commit:
refactor(controller): change userId parameter type from long to Long in the controller

Commit:
refactor: pass GoalFilterDto as a method parameter in controller endpoints

Commit:
refactor: rename methods for consistency in naming conventions
- Renamed findGoalsByUserId to getGetGoalsByUserId
- Renamed getGoalsByUser to findGoalsByUserId
- Renamed getGetGoalsByUserId to getGoalsByUserId

Commit:
feat: add logging to GoalController methods using SLF4J

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/mapper/goal/GoalMapper.java:
```diff
@@ -0,0 +1,45 @@
+    @Mapping(source = "skillsToAchieve", target = "skillIds", qualifiedByName = "mapSkillToSkillsId")
+    GoalDto toGoalDto(Goal goal);
+
+    @Mapping(source = "parentId", target = "parent", qualifiedByName = "mapIdToParent")
```
Комментарий ниже. Суть: ты не можешь из поля ДТО parentId создать и заммапить новой цели сущность родительской цели

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/mapper/goal/GoalMapper.java:
```diff
@@ -0,0 +1,45 @@
+    GoalDto toGoalDto(Goal goal);
+
+    @Mapping(source = "parentId", target = "parent", qualifiedByName = "mapIdToParent")
+    @Mapping(source = "skillIds", target = "skillsToAchieve", qualifiedByName = "mapSkillIdToListSkill")
```
В соответствии с комментарием ниже, маппинг source = "skillIds", target = "skillsToAchieve" у тебя будет в сервисном слое, т.к. тебе нужны сущности Скиллов, которые ты можешь получить только через репозиторий и назначить их сущности Goal, никак иначе. Этот Mapping надо удалить

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/mapper/goal/GoalMapper.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Named("mapIdToParent")
+    default Goal mapIdToParent(Long id) {
+        Goal goal = new Goal();
```
Этот метод не имеет смысла. Здесь ты не сможешь заммапить сущность по id, для этого нужно идти в репозиторий за родительской целью по parentId и сущности Goal назначать полученную из репо по parentId цель

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/mapper/goal/GoalMapper.java:
```diff
@@ -0,0 +1,45 @@
+    @Named("mapSkillIdToListSkill")
+    default List<Skill> mapSkillIdToListSkill(List<Long> skillId) {
+        return skillId.stream().map(id -> {
+            Skill skill = new Skill();
```
Здесь более сложная логика, чем в методе маппинга листа скиллов из сущности в лист айдишек скиллов в ДТО.
Смотри, вместе с запросом на создание Цели приходит список id существующих в базе скиллов. А ты в этом методе создаешь новые скиллы с теми айдишками, которые есть уже у существующих скиллов. Далее ты этот список новых скиллов с существующими айдишками назначаешь сущности Goal и эту сущность сохраняешь в БД, а у скиллов ничего кроме id то и нет. Тебе в лист скиллов у Goal  нужно положить лист сущействующих сущностей скиллов, которые ты получаешь только из БД, никак иначе ты их не получишь

Решение: тебе нужно удалить этот метод из маппера и перенести ее в сервисный слой, в котором ты достанешь у ДТО лист с айдишками скиллов и через репозиторий SkillRepository достанешь сами сущности скиллов и назначишь их сущности Goal и сохранишь уже эту сущность в БД. Просто скласть в сущность Goal список их новых Skill skill = new Skill(); только с айдишками нельзя.

Этот метод проще, чем тот, который описан выше.
default List<Long> mapSkillToSkillsId(List<Skill> skills) {
        return skills.stream().map(Skill::getId).toList();
    }

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+    private final GoalValidator goalValidator;
+    private final SkillRepository skillRepository;
+    private final UserService userService;
+    private final SkillValidator skillValidator;
```
может мой бзик, но я бы сгруппировала репозитории рядышком с репозиториями, мапперы рядом с мапперами, валидаторы с валидаторами))
private final GoalRepository goalRepository;
    private final SkillRepository skillRepository;
    private final GoalMapper goalMapper;
    private final GoalValidator goalValidator;
    private final SkillValidator skillValidator;
    private final UserService userService;
    private final List<GoalFilter> goalFilters;

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+    @Transactional
+    public GoalDto createGoal(Long userId, GoalDto goalDto) {
+        User user = userService.getUserById(userId);
+        goalValidator.validateUserGoalLimit(userId);
```
у тебя по сути очень "тонкие" методы валидации в классе Валидатор, я бы перенесла логику валидации в сервис, потому что иначе сложно читать твой код, приходится постоянно переключаться между классами, чтобы посмотреть что каждый метод валидирует

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+        return filterGoals(goals, filterDto);
+    }
+
+    private Goal getGoalById(long goalId) {
```
почему ты его сделал приватным? для пользователя в UserService аналогичный метод помечен как публичный, да и в дальнейшем ничего не мешает добавать в контроллере эндпоинт по получению цели по id

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+                .orElseThrow(() -> new EntityNotFoundException("Goal with this id does not exist in the database"));
+    }
+
+    private Goal createGoalEntity(GoalDto goalDto, User user) {
```
у тебя этот метод занимается сохранением сущности в БД, а название говорит о создании сущности цели. Я не знаю с какой целью ты дублируешь методы делая 1 публичный, другой приватный. Это делается в 1 методе: валидация, маппинг в сущность, сохранение сущности в БД через репо. В то же время у тебя запутывающий нейминг методов.

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+        goal.getUsers().add(user);
+        setSkillsToGoal(goalDto.getSkillIds(), goal);
+
+        goalRepository.save(goal);
```
два раза сохранение сущности в БД, сначала через create, потом через save - это нагрузка на БД

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,131 @@
+    private void setSkillsToGoal(List<Long> skillIds, Goal goal) {
+        if (skillIds != null) {
+            skillValidator.validateAllSkillsIdsExist(skillIds);
+            List<Skill> skills = skillRepository.findAllById(skillIds);
```
почему здесь используешь для поиска всех скиллов по id метод репозитория, а для поиска пользователя по id, ты создаешь в userService метод для этого и тянешь из сервиса, а для скилла ты решил идти сразу в репо не создавая сервис? Если ты решил ходить в БД за "чужой" сушностью, то тебе надо выбрать какой-то 1 подход и следовать ему.

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/user/UserService.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Transactional
+    public User getUserById(Long userId) {
+        return userRepository.findById(userId)
```
для Goal ты создал отдельный класс с валидацией есть ли такая цель в БД по id, а почему тогда не создал для поиска пользователя по id аналогичный валидатор? В твоей случае это нарушает единообразие кода, если ты для Цели вынес это в валидатор, а для пользователя нет, то код становится нестандартизированным, а беспорядочным

Commit:
fix: changes GoalController and GoalDto adding Validation for Dto

Commit:
refactor: upgrade createGoal method

Commit:
feat: add validation for skill IDs in createGoal method
feat: add mentor validation to ensure association with user

Commit:
refactor: add goalToUpdate method in GoalService

Commit:
refactor: add deleteToGoal method in GoalService

Commit:
refactor: commit some code lines

Commit:
fix: add new logic maxActiveGoalsCount

Commit:
refactor: some useless methods

Commit:
refactor: delete some useless class

Commit:
refactor: extract duplicate code into helper methods in GoalService

Commit:
refactor: delete exception

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/filter/goal/GoalStatusFilter.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public Stream<Goal> apply(Stream<Goal> goals, GoalFilterDto filter) {
+        return goals.filter(goal -> goal.getStatus() == filter.getStatus());
```
статус это стринга, для нее ты не моешь использовать оператор ==, только equals

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -17,7 +17,7 @@ public interface GoalRepository extends JpaRepository<Goal, Long> {
             WHERE ug.user_id = ?1
             """)
-    Stream<Goal> findGoalsByUserId(long userId);
+    Stream<Goal> getGoalsByUserIdId(long userId);
```
для чего ты переименовал метод репозитория?))

EkaterinaBadratdinova commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,156 @@
+                .toList();
+    }
+
+    // Вспомогательные методы
```
не оставляем комментарии в коде

Commit:
feat: add global exception handler with detailed error responses and logging

Commit:
refactor: enhance GoalService with validations, exception handling, and logging

Commit:
refactor: use equals for status comparison in Goal filter application

zaqqaz226 commented on src/main/java/school/faang/user_service/exception/handler/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class ErrorResponse {
```
Использование обертки для ответа с ошибкой - отлично!

zaqqaz226 commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,172 @@
+
+    @Transactional
+    public GoalDto createGoal(Long userId, GoalDto goalDto) {
+        User user = findUserById(userId);
```
Обратись тут спокойно к репозиторию. По сути у тебя происходит вызов метода, который просто вызывает метод, те получается излишне сложно.

InfernalBeard replied:
У меня этот метод с такой логикой работы
    private User findUserById(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> {
                    log.error("User with ID {} not found", userId);
                    return new ResourceNotFoundException("User not found by Id: " + userId);
                });
    }

zaqqaz226 replied:
Да, его логику можно просто перенести сюда. Те нету смысла выносить это в отдельным метод.

InfernalBeard replied:
Исправил!

Commit:
refactor: delete @Transaction

Commit:
refactor

InfernalBeard committed:
Merge branch 'Titan-master-stream7' into Titan-master-stream7-adskaiaboroda-BJS2-40568

zaqqaz226 commented:
Приступай к тестам

Commit:
refactor method findUserById
